File: src/206.reverse-linked-list.ts
// Given the head of a singly linked list, reverse the list, and return the reversed list.
function reverseList(head: ListNode | null): ListNode | null {
  // O(n)
  let prev: ListNode | null = null,
    current: ListNode | null = head;
  while (current !== null) {
    const nextTemp: ListNode | null = current.next;
    current.next = prev;
    prev = current;
    current = nextTemp;
  }
  return prev;
}

File: src/python/206.reverse-linked-list.py
# Given the head of a singly linked list, reverse the list, and return the reversed list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
    # Time: O(n)
    # Space: O(1)
    if not head or not head.next:
        return head
    prev = None
    curr = head
    while curr:
        next_node = curr.next
        curr.next = prev
        prev = curr
        curr = next_node
    return prev
